File: bot.js
const fs = require("fs");
const https = require("https");

const API = "https://www.nationstates.net/cgi-bin/api.cgi";

const TGID = process.env.TGID;
const SECRET_KEY = process.env.SECRET_KEY;
const CLIENT_KEY = process.env.NS_CLIENT_KEY;

const USER_AGENT = "Slavstonia AutoTelegram Bot";
const STATE_FILE = "state.json";

const SEND_DELAY = 180000; // 3 minutes
const RETRIES = 3;

if (!TGID || !SECRET_KEY || !CLIENT_KEY) {
    console.error("Missing GitHub Secret.");
    process.exit(1);
}

function sleep(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
}

function loadState() {
    let state = {};

    if (fs.existsSync(STATE_FILE)) {
        try {
            state = JSON.parse(
                fs.readFileSync(STATE_FILE, "utf8")
            );
        } catch {
            console.log("Creating fresh state.");
        }
    }

    state.initialized = state.initialized ?? false;
    state.lastHappeningID = Number(state.lastHappeningID || 0);
    state.seenNew = Array.isArray(state.seenNew)
        ? state.seenNew
        : [];
    state.processed = Array.isArray(state.processed)
        ? state.processed
        : [];
    state.queue = Array.isArray(state.queue)
        ? state.queue
        : [];
    state.lastSend = Number(state.lastSend || 0);

    return state;
}

function saveState(state) {
    fs.writeFileSync(
        STATE_FILE,
        JSON.stringify(state, null, 2)
    );
}

function request(params) {
    return new Promise((resolve, reject) => {
        const url = new URL(API);

        for (const [key, value] of Object.entries(params)) {
            url.searchParams.set(key, value);
        }

        const req = https.get(
            url,
            {
                headers: {
                    "User-Agent": USER_AGENT
                }
            },
            res => {
                let data = "";

                res.on("data", chunk => {
                    data += chunk;
                });

                res.on("end", () => {
                    if (res.statusCode >= 400) {
                        const error = new Error(
                            `HTTP ${res.statusCode}: ${data}`
                        );

                        error.statusCode = res.statusCode;
                        reject(error);
                        return;
                    }

                    resolve(data);
                });
            }
        );

        req.setTimeout(30000, () => {
            req.destroy(
                new Error("Request timed out")
            );
        });

        req.on("error", reject);
    });
}

/* =========================
   NEWEST NATIONS
========================= */

async function getNewestNations() {
    const xml = await request({
        q: "newnations"
    });

    const match = xml.match(
        /<NEWNATIONS>([\s\S]*?)<\/NEWNATIONS>/i
    );

    if (!match) {
        console.log(
            "Could not find NEWNATIONS in API response."
        );

        console.log(xml);

        return [];
    }

    const nations = match[1]
        .split(",")
        .map(nation =>
            nation
                .trim()
                .replace(/^@@|@@$/g, "")
                .toLowerCase()
        )
        .filter(Boolean);

    return [...new Set(nations)].slice(0, 50);
}

/* =========================
   NEW WA NATIONS
========================= */

async function getNewWANations(state) {
    const xml = await request({
        q: "happenings",
        filter: "member",
        limit: "100",
        sinceid: String(state.lastHappeningID)
    });

    const nations = [];

    const eventRegex =
        /<EVENT[^>]*id="(\d+)"[^>]*>[\s\S]*?<TEXT>([\s\S]*?)<\/TEXT>[\s\S]*?<\/EVENT>/gi;

    let match;
    let highestID = state.lastHappeningID;

    while ((match = eventRegex.exec(xml)) !== null) {
        const id = Number(match[1]);

        if (id > highestID) {
            highestID = id;
        }

        const text = match[2]
            .replace(/<!\[CDATA\[|\]\]>/g, "")
            .replace(/<[^>]+>/g, "")
            .trim();

        const admitted = text.match(
            /(.+?) was admitted to the World Assembly\./i
        );

        if (admitted) {
            const nation = admitted[1]
                .trim()
                .replace(/^@@|@@$/g, "")
                .toLowerCase();

            if (nation) {
                nations.push(nation);
            }
        }
    }

    state.lastHappeningID = highestID;

    return [...new Set(nations)];
}

/* =========================
   QUEUE
========================= */

function addTarget(state, nation, reason) {
    nation = nation
        .toLowerCase()
        .replace(/^@@|@@$/g, "");

    if (!nation) return;

    if (state.processed.includes(nation)) {
        return;
    }

    if (
        state.queue.some(
            x => x.nation === nation
        )
    ) {
        return;
    }

    state.queue.push({
        nation,
        reason
    });

    console.log(
        `QUEUED: ${nation} (${reason})`
    );
}

/* =========================
   SEND TELEGRAM
========================= */

async function sendTG(nation) {
    console.log("");
    console.log(
        `========== SENDING TO ${nation} ==========`
    );

    for (
        let attempt = 1;
        attempt <= RETRIES;
        attempt++
    ) {
        try {
            const response = await request({
                a: "sendTG",
                client: CLIENT_KEY,
                tgid: TGID,
                key: SECRET_KEY,
                to: nation
            });

            console.log(
                "NationStates API response:"
            );

            console.log(response);

            if (/error/i.test(response)) {
                throw new Error(
                    `NationStates API error: ${response}`
                );
            }

            console.log(
                `SEND ACCEPTED: ${nation}`
            );

            return true;

        } catch (error) {
            console.log(
                `Attempt ${attempt}/${RETRIES} failed: ${error.message}`
            );

            if (attempt < RETRIES) {
                console.log(
                    "Waiting 60 seconds before retry..."
                );

                await sleep(60000);
            }
        }
    }

    return false;
}

/* =========================
   DISCOVER TARGETS
========================= */

async function discover(state) {

    console.log("");
    console.log(
        "===== NEWEST NATIONS ====="
    );

    const newest =
        await getNewestNations();

    console.log(
        `Found ${newest.length} newest nations.`
    );

    for (const nation of newest) {
        addTarget(
            state,
            nation,
            "newest nation"
        );
    }

    console.log("");
    console.log(
        "===== NEW WA NATIONS ====="
    );

    const newWA =
        await getNewWANations(state);

    console.log(
        `Found ${newWA.length} new WA nations.`
    );

    for (const nation of newWA) {
        addTarget(
            state,
            nation,
            "new WA nation"
        );
    }

    state.seenNew = newest;
    state.initialized = true;

    saveState(state);
}

/* =========================
   PROCESS QUEUE
========================= */

async function processQueue(state) {

    console.log("");

    console.log(
        `===== ${state.queue.length} TARGETS IN QUEUE =====`
    );

    while (state.queue.length > 0) {

        const target =
            state.queue[0];

        if (state.lastSend !== 0) {

            const elapsed =
                Date.now() -
                state.lastSend;

            if (elapsed < SEND_DELAY) {

                const remaining =
                    SEND_DELAY -
                    elapsed;

                console.log(
                    `Waiting ${Math.ceil(
                        remaining / 1000
                    )} seconds...`
                );

                await sleep(
                    remaining
                );
            }
        }

        const success =
            await sendTG(
                target.nation
            );

        if (!success) {

            console.log(
                `Keeping ${target.nation} in queue for the next run.`
            );

            saveState(state);

            process.exit(1);
        }

        state.queue.shift();

        if (
            !state.processed.includes(
                target.nation
            )
        ) {
            state.processed.push(
                target.nation
            );
        }

        state.lastSend =
            Date.now();

        saveState(state);

        console.log(
            `DONE: ${target.nation}`
        );
    }

    console.log("");
    console.log(
        "===== QUEUE EMPTY ====="
    );
}

/* =========================
   MAIN
========================= */

async function main() {

    console.log("");

    console.log(
        "===== SLAVSTONIA AUTOTELEGRAM ====="
    );

    const state =
        loadState();

    await discover(state);

    await processQueue(state);

    saveState(state);

    console.log(
        "===== RUN FINISHED ====="
    );
}

main().catch(error => {

    console.error(
        "FATAL ERROR:",
        error
    );

    process.exit(1);
});

File: bot.py
import json
import os
import time
import xml.etree.ElementTree as ET
from pathlib import Path

import requests
from dotenv import load_dotenv

API_URL = "https://www.nationstates.net/cgi-bin/api.cgi"
STATE_FILE = Path("state.json")

load_dotenv()

CLIENT_KEY = os.getenv("NS_CLIENT_KEY", "").strip()

NEW_NATIONS_TGID = os.getenv("NEW_NATIONS_TGID", "").strip()
NEW_NATIONS_SECRET = os.getenv("NEW_NATIONS_SECRET", "").strip()

NEW_WA_TGID = os.getenv("NEW_WA_TGID", "").strip()
NEW_WA_SECRET = os.getenv("NEW_WA_SECRET", "").strip()

USER_AGENT = os.getenv(
    "USER_AGENT",
    "NationStates AutoTelegram/1.0 (contact: replace-me@example.com)"
).strip()

POLL_SECONDS = max(60, int(os.getenv("POLL_SECONDS", "180")))

# Recruitment TGs must be spaced by at least 180 seconds.
TG_DELAY = 180

session = requests.Session()
session.headers.update({
    "User-Agent": USER_AGENT,
    "Accept": "application/xml,text/xml;q=0.9,*/*;q=0.8",
})


def check_settings():
    missing = []

    if not CLIENT_KEY:
        missing.append("NS_CLIENT_KEY")

    if not NEW_NATIONS_TGID or not NEW_NATIONS_SECRET:
        missing.append("NEW_NATIONS_TGID / NEW_NATIONS_SECRET")

    if not NEW_WA_TGID or not NEW_WA_SECRET:
        missing.append("NEW_WA_TGID / NEW_WA_SECRET")

    if not USER_AGENT or "replace-me@example.com" in USER_AGENT:
        missing.append("USER_AGENT")

    if missing:
        print("Missing settings:")
        for item in missing:
            print("-", item)
        raise SystemExit


def load_state():
    if not STATE_FILE.exists():
        return {
            "initialized": False,
            "seen_new_nations": [],
            "seen_wa_members": [],
            "queue": [],
            "last_tg": 0
        }

    try:
        return json.loads(
            STATE_FILE.read_text(encoding="utf-8")
        )
    except Exception:
        return {
            "initialized": False,
            "seen_new_nations": [],
            "seen_wa_members": [],
            "queue": [],
            "last_tg": 0
        }


def save_state(state):
    STATE_FILE.write_text(
        json.dumps(state, indent=2),
        encoding="utf-8"
    )


def api_request(params):
    response = session.get(
        API_URL,
        params=params,
        timeout=45
    )

    if response.status_code == 429:
        raise RuntimeError("NationStates API rate limit reached.")

    response.raise_for_status()

    return ET.fromstring(response.content)


def get_new_nations():
    root = api_request({
        "q": "newnations"
    })

    nations = []

    for element in root.iter():
        if element.tag.upper() == "NATION":
            name = (
                element.text
                or element.attrib.get("name", "")
            ).strip()

            if name:
                nations.append(name.lower())

    return list(dict.fromkeys(nations))


def get_wa_members():
    root = api_request({
        "wa": "1",
        "q": "members"
    })

    members = set()

    for element in root.iter():
        if element.tag.upper() in ("NATION", "MEMBER"):
            name = (
                element.text
                or element.attrib.get("name", "")
                or element.attrib.get("nation", "")
            ).strip()

            if name:
                members.add(name.lower())

    return members


def add_to_queue(state, nation, target_type):
    nation = nation.lower().strip()

    for item in state["queue"]:
        if (
            item["nation"] == nation
            and item["type"] == target_type
        ):
            return

    state["queue"].append({
        "nation": nation,
        "type": target_type
    })


def scan(state):
    print("Checking NationStates...")

    new_nations = get_new_nations()
    wa_members = get_wa_members()

    old_new_nations = set(
        state["seen_new_nations"]
    )

    old_wa_members = set(
        state["seen_wa_members"]
    )

    # First run: record existing nations instead of
    # immediately messaging all of them.
    if not state["initialized"]:

        state["seen_new_nations"] = list(
            new_nations
        )

        state["seen_wa_members"] = list(
            wa_members
        )

        state["initialized"] = True

        save_state(state)

        print("First scan complete.")
        print("Existing targets were recorded.")
        return

    # Newly founded nations
    for nation in new_nations:

        if nation not in old_new_nations:
            add_to_queue(
                state,
                nation,
                "new_nation"
            )

    # Newly detected WA members
    for nation in wa_members:

        if nation not in old_wa_members:
            add_to_queue(
                state,
                nation,
                "new_wa"
            )

    state["seen_new_nations"] = list(
        new_nations
    )

    state["seen_wa_members"] = list(
        wa_members
    )

    save_state(state)

    print(
        "Queue:",
        len(state["queue"])
    )


def send_telegram(item):
    nation = item["nation"]
    target_type = item["type"]

    if target_type == "new_nation":

        tgid = NEW_NATIONS_TGID
        secret = NEW_NATIONS_SECRET

    else:

        tgid = NEW_WA_TGID
        secret = NEW_WA_SECRET

    params = {
        "a": "sendTG",
        "client": CLIENT_KEY,
        "tgid": tgid,
        "key": secret,
        "to": nation
    }

    response = session.get(
        API_URL,
        params=params,
        timeout=45
    )

    if response.status_code == 429:
        print("Telegram rate limit reached.")
        return False

    if response.status_code >= 400:
        print(
            "Telegram failed:",
            nation,
            response.status_code
        )
        return False

    print(
        "Telegram sent to:",
        nation
    )

    return True


def process_queue(state):

    if not state["queue"]:
        return

    elapsed = (
        time.time()
        - float(state.get("last_tg", 0))
    )

    if elapsed < TG_DELAY:

        remaining = int(
            TG_DELAY - elapsed
        )

        print(
            "Waiting",
            remaining,
            "seconds before next TG."
        )

        return

    item = state["queue"][0]

    if send_telegram(item):

        state["queue"].pop(0)

        state["last_tg"] = time.time()

        save_state(state)


def main():

    check_settings()

    print("--------------------------------")
    print("NationStates AutoTelegram")
    print("--------------------------------")
    print("Recruitment TG delay:", TG_DELAY)
    print("Polling every:", POLL_SECONDS)
    print()

    state = load_state()

    while True:

        try:

            scan(state)

            process_queue(state)

            save_state(state)

        except KeyboardInterrupt:

            print("Bot stopped.")
            break

        except Exception as error:

            print(
                "ERROR:",
                error
            )

        print(
            "Sleeping",
            POLL_SECONDS,
            "seconds..."
        )

        time.sleep(POLL_SECONDS)


if __name__ == "__main__":
    main()
